Query generator functions for multivalued property searches

AddisonDunn opened the pull request:
## Product Description

There is no user interface in this PR. It implements part of the back-end changes needed to allow searching on multivalued properties (which are generated by a multi-select question).

Here is [the related Jira ticket](https://dimagi-dev.atlassian.net/browse/USH-1731?atlOrigin=eyJpIjoiOTBiZjViNzk1ODgzNDM1N2JiMWUxNGU1MDRkY2NiZmEiLCJwIjoiaiJ9).

[Related spec created by Simon](https://docs.google.com/document/d/1fxb7kUw-GtCitJJIA_vTs8hZVKb7oqCRoMmpyQeRobY/edit?usp=sharing). 

The back-end changes for adding multivalued property searching include app model, suite XML, and XPath processing changes. This PR covers just the XPath processing changes: three functions that generate ES queries from XPath.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

[Simon & Dominic's PR](https://github.com/dimagi/commcare-hq/pull/31115) related to subcase searching edits some of the same files but it looks like different lines of code for the most part. So there may or may not be merge conflicts. Even though there are no user-facing changes I will deploy this branch to staging soon so that we can get ahead of possible merge conflicts (per Jenny's request this week).

Per the code, this PR works under the assumption that the passed XPath will have the exact syntax: `selected(name, 'value')`. I believe I parsed out the function correctly in my change to `filter_dsl.py`, but let me know if it doesn't look right.

One other note: I for the most part followed the spec word for word but I found one more change necessary, and that's the addition of an operator to the `match` function. 

Best reviewed by commit.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

None. This code should be inaccessible until UI code is added.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

Passed all tests in the `case_search` app.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Added tests.

### QA Plan

None, this is all back-end and should have no user impact.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Commit:
add multivalue mode and operator args to case query methods

Commit:
add new selected XPath functions

Commit:
add tests

joxl commented on corehq/apps/es/queries.py:
```diff
@@ -93,11 +93,13 @@ def ids_query(doc_ids):
             field: {
                 "query": search_string,
+                # OR is the accepted default for the operator on an ES match query
+                "operator": 'and' if operator == 'and' else 'or',
```
This function should also validate the value of `operator`.

```python
from ..queries import match

# <snip>

    def test_match_raises_with_invalid_operator(self):
        with self.assertRaises(ValueError):
            query = match("cyrus", "pet_name", operator="And")
```

snopoke commented on corehq/apps/case_search/filter_dsl.py:
```diff
@@ -213,6 +213,13 @@ def _comparison(node):
+
+        if isinstance(node, FunctionCall):
+            if node.name in XPATH_FUNCTIONS:
+                return XPATH_FUNCTIONS[node.name](node, fuzzy)
```
We can't support all functions here, only 'query' function: `date('2022-01-01')` is not a valid query.

We should probably separate value functions from query functions since the new functions in this PR don't make sense as value functions:
```
# good
dob = date('2019-01-01')

# bad
dob = selected(prop, 'value')
```

snopoke replied:
I ended up doing this in the subcase PR: https://github.com/dimagi/commcare-hq/pull/31115/files#diff-466c8903c1e7201de8bf4c76a3abd642f389e285d8bcf27f4509c834b0aae580R224-R231

Feel free to copy it over, I'll handle any conflicts on my PR.

snopoke commented on corehq/apps/case_search/tests/test_filter_dsl.py:
```diff
@@ -344,6 +344,165 @@ def test_nested_filter(self):
 
+    def test_selected(self):
+        parsed = parse_xpath("selected(first_name, 'Jon')")
+        expected_filter_single = {
```
This query is incorrect. `selected` should match if the property contains the term:

```
first_name = 'Jon James'   # match
first_name = 'Mick Jon'   # match
first_name = 'Mick Jo'   # no match (unless also using fuzzy)
first_name = 'Mick'   # no match
```

AddisonDunn replied:
Then in that case the query for correct `selected` should be identical to `selected-any` and look like this?

```
"must": {
  "match": {
    "case_properties.value": {
       "query": "Jon",
       "operator": "or",
       "fuzziness": "0"
     }
  }
}
```

Regardless of how many terms we're searching on?

snopoke replied:
that's correct

AddisonDunn replied:
👍  corresponding changes made with recent commits

snopoke commented on corehq/apps/case_search/xpath_functions.py:
```diff
@@ -35,6 +38,36 @@ def date(node):
+    search_values = node.args[1]
+    search_mode = None
+    if 1 < len(search_values.split(' ')):
+        search_mode = 'OR'  # This will actually make the query the same as with selected_any
```
I don't think this should be conditional on the number of terms

snopoke commented on corehq/apps/case_search/xpath_functions.py:
```diff
@@ -35,6 +38,36 @@ def date(node):
+    return case_property_query(property_name, search_values, fuzzy=fuzzy, mode='OR')
+
+
+def selected_all(node, fuzzy):
+    if len(node.args) != 2:
+        raise XPathFunctionException(_("The \"selected-all\" function accepts exactly two arguments."))
+    property_name = serialize(node.args[0])
+    search_values = node.args[1]
+    return case_property_query(property_name, search_values, fuzzy=fuzzy, mode='AND')
```
You could make all of these call a common function:
```suggestion
def selected_all(node, fuzzy):
    return _selected_query(node, fuzzy, 'AND')


def _selected_query(node, fuzzy, operator):
    if len(node.args) != 2:
        raise XPathFunctionException(_("The \"{name}\" function accepts exactly two arguments.").format(name=node.name))
    property_name = serialize(node.args[0])
    search_values = node.args[1]
    return case_property_query(property_name, search_values, fuzzy=fuzzy, mode=operator)
```

snopoke commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,32 @@ def case_property_filter(case_property_name, value):
             case_property_text_query(case_property_name, value),
         )
+    if fuzzy and mode == 'AND':
+        return filters.AND(
```
This should still be an OR. We want: "does it match exactly OR does it fuzzy match"

AddisonDunn replied:
I see--fixed

snopoke commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,32 @@ def case_property_filter(case_property_name, value):
+    if fuzzy and mode == 'AND':
+        return filters.AND(
+            case_property_text_query(case_property_name, value, fuzziness='AUTO', operator='and'),
+            case_property_text_query(case_property_name, value),
```
the operator needs to be passed here too

AddisonDunn replied:
Yeah I was trying to follow the spec too closely. Done

snopoke commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,32 @@ def case_property_filter(case_property_name, value):
     """
     if value is None:
         raise TypeError("You cannot pass 'None' as a case property value")
+    if mode not in ['AND', 'OR', None]:
```
if you make these lower case (maybe use constants too) you could just pass it through to `case_property_text_query`

Commit:
fix broken test in ES app

Commit:
split out XPATH_FUNCTIONS

Commit:
use common function for query functions

Commit:
validate operator in match function

Commit:
update test for updated selected query function

Commit:
pass mode through to operator

snopoke commented on corehq/apps/case_search/xpath_functions.py:
```diff
@@ -35,6 +38,32 @@ def date(node):
 
 
-XPATH_FUNCTIONS = {
+def _selected_query(node, fuzzy, operator):
```
nit: this function should after the other functions to follow [Newspaper Metaphor](https://github.com/jnguyen095/clean-code/issues/28)

snopoke commented on corehq/apps/case_search/xpath_functions.py:
```diff
@@ -35,6 +38,32 @@ def date(node):
+    return _selected_query(node, fuzzy=fuzzy, operator='or')
+
+
+def selected_any(node, fuzzy):
```
since this is identical to `selected` you could delete it and alias `selected` to `selected-any` in the function map

snopoke commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,25 @@ def case_property_filter(case_property_name, value):
-            case_property_text_query(case_property_name, value),
+            case_property_text_query(case_property_name, value, operator=mode),
         )
+    if not fuzzy and mode in ['or', 'and']:
```
nit: the fuzzy check is redundant since we've already passed that check above

```suggestion
    if mode in ['or', 'and']:
```

AddisonDunn replied:
I left it the way it is because it feels clearer to me

snopoke commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,25 @@ def case_property_filter(case_property_name, value):
 
 
-def case_property_query(case_property_name, value, fuzzy=False):
+def case_property_query(case_property_name, value, fuzzy=False, mode=None):
```
nit: `mode` could be more descriptive. Maybe `multivalue_mode`?

snopoke reviewed (approved):
some small comments but otherwise looks good

Commit:
refactoring

AddisonDunn commented:
Thanks @snopoke , 2nd round of changes made!

snopoke reviewed (approved):
I'm approving this but I think it would be good to add a test which actually verifies that the queries do what we expect.

The [test_case_search_filters.py](https://github.com/dimagi/commcare-hq/blob/7ca190bf277c833dae34a706d9985877514b1504/corehq/apps/case_search/tests/test_case_search_filters.py#L10) file has tests that do this for other queries. I think 2 tests should be sufficient, one for `any` and one for `and`

esoergel commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,25 @@ def case_property_filter(case_property_name, value):
+    if multivalue_mode not in ['and', 'or', None]:
+        raise ValueError(" 'mode' must be one of: 'and', 'or', None")
     if value == '':
         return case_property_missing(case_property_name)
```
Does multivalue_mode interact with this branch at all?  Ie, what should
`selected-all(my_prop, '')` and `selected-any(my_prop, '')` return?  As is, those will both match only cases where the user selected nothing.

AddisonDunn replied:
Could you clarify what you're asking a bit more for me? The value of `multivalue_mode` and the cases where a `selected` method is passed an empty string seem unrelated.

snopoke replied:
In terms of how mobile handles this, it will return `True` if the case property contains a double space (or is empty):
```java
(" " + property_value + " ").contains(" " + input + " ")
```

It's not exactly obvious to me what the case search functionality should be.

snopoke replied:
I think the options are to either:
1. reject queries with empty input
2. treat it as a 'missing property' query
3. do the actual multivalue query which will return no results

My intuition is that 1 is the most correct since 2 or 3 yield surprising results for a 'selected' query.

(and there isn't really a way to make this function the same as mobile without doing a regex query which I would say is out of scope).

esoergel replied:
Waiting on feedback from AEs sounds good to me.  Definitely not a fan of implementing the mobile double-space behavior, since that doesn't sound deliberate.

> Could you clarify what you're asking a bit more for me? The value of `multivalue_mode` and the cases where a selected method is passed an empty string seem unrelated.

Simon elaborated a bit on what I meant, which is basically that there isn't an obvious correct behavior for what this function should do when using a multivalue mode and the value is the empty string.

For single select properties, it makes sense - if you say `favorite_color = 'purple'`, then it matches only those where the field is exactly that.  If you say `favorite_color = ''`, then you match only those where favorite_color is unspecified.

However, with multiselect it's more complicated.  If the caller says `selected-any("favorite_colors", "purple blue")`, then this should match cases where favorite_colors is "purple" or "red blue", but not "green yellow".  That much is straightforward.

But what if they say `selected-any("favorite_colors", "")`?  Does that match only users that haven't specified favorite_colors (the current behavior)?  All users?  None of them?  What about `selected-all`?

As an analogue, consider python.  This behavior seems intuitive to me, and may be a model we could follow here:
```python
In [1]: all(n in [1, 2, 3] for n in [])
Out[1]: True

In [2]: any(n in [1, 2, 3] for n in [])
Out[2]: False
```

AddisonDunn replied:
Thanks for the clarification, Ethan. I'm on the same page now.

Commit:
add higher level query test

AddisonDunn commented:
> I'm approving this but I think it would be good to add a test which actually verifies that the queries do what we expect.
> 
> The [test_case_search_filters.py](https://github.com/dimagi/commcare-hq/blob/7ca190bf277c833dae34a706d9985877514b1504/corehq/apps/case_search/tests/test_case_search_filters.py#L10) file has tests that do this for other queries. I think 2 tests should be sufficient, one for `any` and one for `and`

Hey @snopoke, I added these tests. While adding them a requirement to "triage" the query properly popped up. I noticed in your spec you suggested a `multivalue_mode` on `SearchCriteria`. But based just on the way the code is written so far, I thought adding a check based on the key, i.e. `x_commcare_{property_name}_config` was more appropriate. Is that the correct key? Let me know if what I've added is accurate/compatible with what the case search request will look like.

snopoke commented on corehq/apps/case_search/tests/test_case_search_filters.py:
```diff
@@ -116,3 +116,42 @@ def test_blank_case_search_parent(self):
         }).get_ids()
         self.assertItemsEqual(actual, ['c4', 'c6', 'c8'])
+
+    def test_multivalued_property_case_search(self):
```
```suggestion
    def test_select_any_function(self):
```

snopoke commented on corehq/apps/case_search/tests/test_case_search_filters.py:
```diff
@@ -116,3 +116,42 @@ def test_blank_case_search_parent(self):
+            ['c1', 'c3']
+        )
+
+        self._create_case_search_config()
```
This should be a separate test:

```suggestion
    def test_select_all_function(self):
        self._create_case_search_config()
```

snopoke commented on corehq/apps/case_search/tests/test_case_search_filters.py:
```diff
@@ -116,3 +116,42 @@ def test_blank_case_search_parent(self):
+            get_case_search_query(
+                self.domain,
+                ['song'],
+                {'x_commcare_description_config': "selected-any(description, 'New York Boston')"},
```
This should be using the `_xpath_query` key (same below):
```suggestion
                {'_xpath_query': "selected-any(description, 'New York Boston')"},
```

Or you could actually use the `CASE_SEARCH_XPATH_QUERY_KEY` constant.

snopoke commented on corehq/apps/case_search/tests/test_case_search_filters.py:
```diff
@@ -116,3 +116,42 @@ def test_blank_case_search_parent(self):
+            get_case_search_query(
+                self.domain,
+                ['song'],
+                {'x_commcare_description_config': "selected-all(description, 'New York Boston')"},
```
```suggestion
                {'_xpath_query': "selected-all(description, 'New York Boston')"},
```

snopoke reviewed (changes requested):
> I noticed in your spec you suggested a multivalue_mode on SearchCriteria. But based just on the way the code is written so far, I thought adding a check based on the key, i.e. x_commcare_{property_name}_config was more appropriate. Is that the correct key? 

Those changes were to facilitate doing multivalue search from normal search properties (without using the xpath query) so you can drop that commit (c35044f85c5ccc368282b75dd54621b0c4719b0f).

I've made a few small recommendations to the tests.

Commit:
test using regular _xpath_query syntax and split out tests

AddisonDunn commented:
@snopoke thanks for the clarification, it makes sense we're just using `_xpath_query` for the key. I've dropped that commit and confirmed the tests work. I've also made the other small changes.

joxl commented on corehq/apps/es/queries.py:
```diff
@@ -93,11 +93,15 @@ def ids_query(doc_ids):
 
 
-def match(search_string, field, fuzziness="AUTO"):
+def match(search_string, field, fuzziness="AUTO", operator=None):
```
Tests for this new functionality should live in `corehq/apps/es/tests/test_queries.py`.

AddisonDunn replied:
Same as my comment above, are you suggesting I add new, more specific tests or that I move existing ones?

joxl replied:
Just moving the existing one is fine. The diff view had me thinking you added more than one test to `corehq/apps/case_search/tests/test_case_search_es_queries.py`, but I see that's not the case. To simplify my ask: everything you put in that test (in commit c06041762f729be5cdcd3c34b00cf76cd15dc7a5) should be in `corehq/apps/es/tests/test_queries.py` instead.

AddisonDunn replied:
Gotcha, thanks for clarifying Joel. Test moved in my most recent commit!

AddisonDunn commented on corehq/apps/es/case_search.py:
```diff
@@ -141,21 +141,25 @@ def case_property_filter(case_property_name, value):
 
 
-def case_property_query(case_property_name, value, fuzzy=False):
+def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None):
```
Hey Joel, this new functionality is tested by extension in `test_filter_dsl.py` with what I've got in [this commit](https://github.com/dimagi/commcare-hq/pull/31187/commits/7dc27805ffd20a7f28899179cd3418c356163d91). The inputted `selected` functions use `case_property_query` with a `multivalue_mode` and check the ES output.

Are you suggesting I add more specific test cases for just this method, or that these tests should be in `test_case_search_es.py` instead?

joxl replied:
> Are you suggesting I add more specific test cases for just this method

Yes, in this situation there should be additional tests in `corehq/apps/es/tests/test_case_search_es.py` to test the keyword arg changes made on the `case_property_query()` and `case_property_text_query()` functions. Your `filter_dsl` tests are fine as is (those tests verify logic that's more specific than just these two functions on their own).

I haven't had a look at the existing tests in `corehq/apps/es/tests/test_case_search_es.py`, but the fact that you were able to modify the behavior of these two functions without causing failures suggests that there probably isn't sufficient testing of these functions yet. If that's the true then it would be nice if you added tests that at least verify the functionality provided by your new keyword arguments (bonus points for adding tests for untested functionality you didn't change :grin:).

Commit:
move match function test

snopoke committed:
pass domain to query functions

snopoke committed:
pass 'filter_part' to exceptions

snopoke committed:
move 'selected-*' functions to query_function.py

snopoke committed:
Merge branch 'master' into ad/multivalued-search-functions

snopoke commented:
@AddisonDunn I've merged in changes from `master` which include the subcase search changes. I've also pushed a few more commits to resolve post merge issues:

* 70e2a58ecc6 move 'selected-*' functions to query_function.py
* d29f739e64e pass 'filter_part' to exceptions
* 8a89bea5139 pass domain to query functions

Commit:
add multivalue_mode validation test for case_property_query
